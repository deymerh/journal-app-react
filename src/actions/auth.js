import { firebase, gogleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';

export const loginWithEmailAndPassword = (email, password) => {
  return (dispatch) => {
    setTimeout(() => {
      dispatch(login(12345, 'Pedro'))
    }, 3500);
  };
};
export const startGoogleLogin = () => {
  return (dispath) => {
    firebase.auth().signInWithPopup(gogleAuthProvider)
      .then(({ user }) => {
        dispath(login(user.uid, user.displayName))
      })
  }
}
export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
});