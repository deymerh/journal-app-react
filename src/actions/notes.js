import Swal from 'sweetalert2';
import { db } from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/journalAdd';
import { types } from '../types/types';

export const startNewNote = ()=>{
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }
    const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
    dispatch(activeNote(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));
  }
};

export const startLoadingNotes = (uid)=>{
  return async (dispatch)=>{
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  }
}

export const activeNote = (id, note)=>({
  type: types.notesActive,
  payload: {
    id,
    ...note
  }
});

export const addNewNote = (id, note)=>({
  type: types.notesAddNew,
  payload: {
    id,
    ...note
  }
});

export const setNotes = (notes)=>({
  type:types.notesLoad,
  payload: notes
});

export const startSaveNote = (note)=>{
  return async(dispatch, getState)=>{
    const {uid} = getState().auth;
    if (!note.url) {
      delete note.url;
    }
    const noteToFirestore = {...note};
    delete noteToFirestore.id;
    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
    dispatch(refreshNote(note));
    Swal.fire('Saved', note.title, 'success');
  }
};

export const refreshNote = (note)=>({
  type: types.notesUpdated,
  payload:note
});

export const startUploadImg = (image)=>{
  return async(dispatch, getState)=>{
    const {active:noteActive} = getState().notes;
    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait...',
      allowOutsideClick: false,
      onBeforeOpen: ()=>{
        Swal.showLoading();
      }
    });
    const { secure_url } = await fileUpload(image);
    noteActive.url = secure_url;
    dispatch(startSaveNote(noteActive));
    Swal.close();
  }
};

export const startDeleteNoteSelected = (id)=>{
  return async(dispatch, getState)=>{
    const {uid} = getState().auth;
    await db.doc(`${uid}/journal/notes/${id}`).delete();
    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id)=>({
  type: types.noteSelectDelete,
  payload: id
});