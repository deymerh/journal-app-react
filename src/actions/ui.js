import { types } from "../types/types"

export const setErrorAction = (err) => ({
  type: types.uiSetError,
  payload: err
});

export const setRemoveErrorAction = () => ({
  type: types.uiRemoveError,
});