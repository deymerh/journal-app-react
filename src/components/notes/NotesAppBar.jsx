import React, { useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploadImg } from '../../actions/notes';

export const NotesAppBar = () => {
  const {active:noteActive} = useSelector(state => state.notes)
  const dispatch = useDispatch();
  const inputFileImg = useRef(null);
  const handleSaveNote = ()=>{
    dispatch(startSaveNote(noteActive));
  }
  const handleUploadPicture = () => {
    inputFileImg.current.click();
    console.log('Click');
  };
  const handleFileChange = (e)=>{
    const image = e.target.files[0];
    if (image) {
      dispatch(startUploadImg(image));
    }
  };
  return (
    <div className="notes__appbar">
      <span>28 de Agosto 2021</span>
      <input
        type="file"
        id="fileSelector"
        name="fileSelector"
        style={{display:'none'}}
        onChange={handleFileChange}
        ref={inputFileImg}
      />
      <div>
        <button
          className="btn"
          onClick={handleUploadPicture}
        >
          Picture
        </button>
        <button className="btn"
          onClick={handleSaveNote}
        >
          Save
        </button>
      </div>
    </div>
  )
}
