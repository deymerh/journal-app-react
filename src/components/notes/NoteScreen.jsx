import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleteNoteSelected } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';

export const NoteScreen = () => {
  const {active:noteActive} = useSelector(state => state.notes);
  const dispatch = useDispatch();
  const [values, handleInputChange, reset] = useForm(noteActive);
  const {title, body} =  values;
  const activeNotedId =  useRef(noteActive.id);
  useEffect(() => {
    if (noteActive.id !== activeNotedId.current) {
      reset(noteActive);
      activeNotedId.current = noteActive.id;
    }
  }, [noteActive, reset]);
  useEffect(() => {
    dispatch(activeNote(noteActive.id, {...values}))
  }, [values, dispatch]);
  const handleDelete = ()=>{
    dispatch(startDeleteNoteSelected(noteActive.id));
  };
  return (
    <div className="notes__main-content animate__animated animate__fadeIn animated__fater">
      <NotesAppBar />
      <div className="notes__content">
        <input
          name="title"
          value={title}
          type="text"
          id=""
          placeholder="Some awesome title"
          className="notes__title-input"
          onChange={handleInputChange}
        />
        <textarea
          name="body"
          value={body}
          id=""
          cols="30"
          rows="10"
          placeholder="what happened today"
          className="notes__textarea"
          onChange={handleInputChange}
        >
        </textarea>
        {
          (noteActive.url) &&
            (
              <div className="notes__image">
                <img src={noteActive.url} alt="Imagen" />
              </div>
            )
        }
      </div>
      <button
        className="btn btn-danger"
        onClick={handleDelete}
      >
        Delete
      </button>
    </div>
  )
}
