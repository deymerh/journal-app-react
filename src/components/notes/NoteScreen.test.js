import '@testing-library/jest-dom';
import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { NoteScreen } from './NoteScreen';
import { activeNote } from '../../actions/notes';

jest.mock('../../actions/notes', ()=>({
  activeNote: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
  auth:{},
  ui:{
    loading: false,
    msgError: null
  },
  notes: {
    active: {
      id: 123456,
      title: 'Hola',
      body: 'mundo',
      date: 0
    },
    notes:  []
  },
}
let store = mockStore(initState); 

describe('Pruebas en el <NoteScreen/>', () => {
  test('Debe mostrarse correctamente', () => {
    store.dispatch = jest.fn()
    const component = render(
      <Provider store={store}>
        <NoteScreen />
      </Provider>
    );
    expect(component).toMatchSnapshot()
  });

  test('Debe disparar el activeNote', async() => {
    store.dispatch = jest.fn()
    const {findByTestId} = render(
      <Provider store={store}>
        <NoteScreen />
      </Provider>
    );
    const titleField = await findByTestId('title');
    const bodyField = await findByTestId('body');
    fireEvent.change(titleField, {target: {value: "Hola"}});
    fireEvent.change(bodyField, {target: {value: "mundo"}});
    expect(activeNote).toHaveBeenLastCalledWith (123456, {"body": "mundo", "date": 0, "id": 123456, "title": "Hola"})
  });
  
});