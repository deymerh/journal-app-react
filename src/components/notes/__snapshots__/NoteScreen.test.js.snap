// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <NoteScreen/> Debe mostrarse correctamente 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="notes__main-content animate__animated animate__fadeIn animated__fater"
      >
        <div
          class="notes__appbar"
        >
          <span>
            28 de Agosto 2021
          </span>
          <input
            id="fileSelector"
            name="fileSelector"
            style="display: none;"
            type="file"
          />
          <div>
            <button
              class="btn"
            >
              Picture
            </button>
            <button
              class="btn"
            >
              Save
            </button>
          </div>
        </div>
        <div
          class="notes__content"
        >
          <input
            class="notes__title-input"
            data-testid="title"
            id=""
            name="title"
            placeholder="Some awesome title"
            type="text"
            value="Hola"
          />
          <textarea
            class="notes__textarea"
            cols="30"
            data-testid="body"
            id=""
            name="body"
            placeholder="what happened today"
            rows="10"
          >
            mundo
          </textarea>
        </div>
        <button
          class="btn btn-danger"
        >
          Delete
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="notes__main-content animate__animated animate__fadeIn animated__fater"
    >
      <div
        class="notes__appbar"
      >
        <span>
          28 de Agosto 2021
        </span>
        <input
          id="fileSelector"
          name="fileSelector"
          style="display: none;"
          type="file"
        />
        <div>
          <button
            class="btn"
          >
            Picture
          </button>
          <button
            class="btn"
          >
            Save
          </button>
        </div>
      </div>
      <div
        class="notes__content"
      >
        <input
          class="notes__title-input"
          data-testid="title"
          id=""
          name="title"
          placeholder="Some awesome title"
          type="text"
          value="Hola"
        />
        <textarea
          class="notes__textarea"
          cols="30"
          data-testid="body"
          id=""
          name="body"
          placeholder="what happened today"
          rows="10"
        >
          mundo
        </textarea>
      </div>
      <button
        class="btn btn-danger"
      >
        Delete
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
